### PR TITLE
[10.0][IMP] medical_prescription_state_verify: Add support for verification procedures

### PR DESCRIPTION
**Pull request by kellylougheed**
- Forward port of #173 
- Automatically moves prescription lines with a verified prescription to the "Hold" stage
- Allows prescription lines to be modified by moving them to the "Hold" stage even if the parent prescription is verified
- If a prescription is verified, the prescription line cannot be modified otherwise (e.g. moved back to "Unverified")

### CODE REVIEW

File: medical_prescription_state/data/base_kanban_stage.xml [hunk 1]
```diff
@@ -36,6 +36,7 @@
         <field name="description">These prescriptions have passed initial verification.</field>
         <field name="sequence" eval="4" />
         <field name="fold" eval="True" />
+        <field name="is_verified" eval="True" />
```
Comment (lasley) [inline medical_prescription_state/data/base_kanban_stage.xml]: This should be in the _verified module

File: medical_prescription_state/models/base_kanban_stage.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+# Copyright 2016-2017 LasLabs Inc.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import fields, models
+
+
+class BaseKanbanStage(models.Model):
+    _inherit = 'base.kanban.stage'
+
+    is_verified = fields.Boolean(
```
Comment (lasley) [inline medical_prescription_state/models/base_kanban_stage.py]: This should be in the _verify module

File: medical_prescription_state_verify/models/medical_prescription_order.py [hunk 3]
```diff
@@ -23,37 +21,18 @@ class MedicalPrescriptionOrder(models.Model):
     _inherit = 'medical.prescription.order'
 
     _ALLOWED_CHANGE_KEYS = ['stage_id', ]
```
Comment (lasley) [inline medical_prescription_state_verify/models/medical_prescription_order.py]: Remove stray `, `

File: medical_prescription_state_verify/models/medical_prescription_order.py [hunk 4]
```diff
@@ -23,37 +21,18 @@ class MedicalPrescriptionOrder(models.Model):
...
-        default=lambda self: self.env.ref(
-            'medical_prescription_state_verify.state_verification'
-        ),
-    )
-
-    state_type = fields.Selection(
-        related='stage_id.type',
-        help="The state type for the order",
-    )
+    _ALLOWED_CHANGE_STATES = ['Verified', 'Cancelled', 'Exception', ]
```
Comment (lasley) [inline medical_prescription_state_verify/models/medical_prescription_order.py]: Remove stray `, `

File: medical_prescription_state_verify/models/medical_prescription_order.py [hunk 5]
```diff
@@ -23,37 +21,18 @@ class MedicalPrescriptionOrder(models.Model):
...
     @api.multi
     def write(self, vals, ):
         """
         Overload write & perform audit validations
-
         Raises:
             ValidationError: When a write is not allowed due to being in a
                 protected state
         """
         for rec_id in self:
```
Comment (lasley) [inline medical_prescription_state_verify/models/medical_prescription_order.py]: `self.filtered(lambda r: r.stage_id.is_verified)` & remove below conditional

File: medical_prescription_state_verify/models/medical_prescription_order.py [hunk 6]
```diff
@@ -71,13 +50,21 @@ def write(self, vals, ):
...
-                    if stage_id.type not in self._ALLOWED_CHANGE_STATES:
+                    if stage_id.name not in self._ALLOWED_CHANGE_STATES:
                         raise exceptions.ValidationError(_(
                             'You cannot move an Rx into this state after it'
                             ' has been verified. [%s]' % rec_id.name
                         ))
 
-            return super(MedicalPrescriptionOrder, self).write(vals)
+            super(MedicalPrescriptionOrder, self).write(vals)
+            if rec_id.stage_id.is_verified:
```
Comment (lasley) [inline medical_prescription_state_verify/models/medical_prescription_order.py:63]: This is outside of the record loop, but using the record. It's also causing two db writes & performing the same conditional twice. All the logic should be moved above & injected into vals before write.

File: medical_prescription_state_verify/models/medical_prescription_order.py [hunk 7]
```diff
@@ -71,13 +50,21 @@ def write(self, vals, ):
...
+                stage_id = self.env['base.kanban.stage'].search(
+                    [("name", "=", "Hold")],
+                    limit=1
+                )
+                for prescription_order_line in \
```
Comment (lasley) [inline medical_prescription_state_verify/models/medical_prescription_order.py:68]: `\` is nasty hack.

File: medical_prescription_state_verify/models/medical_prescription_order_line.py [hunk 8]
```diff
@@ -14,11 +15,7 @@ class MedicalPrescriptionOrderLine(models.Model):
     """
 
     _inherit = 'medical.prescription.order.line'
-
-    state_type = fields.Selection(
-        related='prescription_order_id.state_type',
-        help="The state type for the order"
-    )
+    _name = 'medical.prescription.order.line'
```
Comment (lasley) [inline medical_prescription_state_verify/models/medical_prescription_order_line.py]: This is superfluous unless you are inheriting multiple models at once.

File: medical_prescription_state_verify/models/medical_prescription_order_line.py [hunk 9]
```diff
@@ -29,13 +26,19 @@ def write(self, vals, ):
...
-            if rec_id.state_type == 'verified':
-                raise exceptions.ValidationError(_(
-                    'You cannot edit this value after its parent Rx has'
-                    ' been verified. Please either cancel it, or mark it as'
-                    ' an exception if manual reversals are required. [%s]' %
-                    rec_id.prescription_order_id.name
-                ))
-
-            return super(MedicalPrescriptionOrderLine, self).write(vals)
+        if self.prescription_order_id.stage_id.is_verified:
```
Comment (lasley) [inline medical_prescription_state_verify/models/medical_prescription_order_line.py]: Expects singleton. Write requires a record loop

File: medical_prescription_state_verify/models/medical_prescription_order_line.py [hunk 10]
```diff
@@ -29,13 +26,19 @@ def write(self, vals, ):
...
+            if vals.get('stage_id'):
+                stage_id = self.env['base.kanban.stage'].browse(
+                    vals['stage_id']
+                )
+                if stage_id.name == "Hold":
```
Comment (lasley) [inline medical_prescription_state_verify/models/medical_prescription_order_line.py:34]: Using the name of a stage will not be future proof to user edits. There should be an option on the stage to allow it to be moved to after verification.

File: medical_prescription_state_verify/models/medical_prescription_order_line.py [hunk 11]
```diff
@@ -29,13 +26,19 @@ def write(self, vals, ):
...
+                    return super(
```
Comment (lasley) [inline medical_prescription_state_verify/models/medical_prescription_order_line.py:35]: This logic nesting is too complex.

File: medical_prescription_state_verify/models/medical_prescription_order_line.py [hunk 12]
```diff
@@ -29,13 +26,19 @@ def write(self, vals, ):
...
+                        MedicalPrescriptionOrderLine, self
+                    ).write(vals)
+            raise ValidationError(_(
+                'You cannot edit this value after its parent Rx has'
+                ' been verified. Please either cancel it, or mark it as'
+                ' an exception if manual reversals are required. [%s]' %
+                self.prescription_order_id.name
```
Comment (lasley) [inline medical_prescription_state_verify/models/medical_prescription_order_line.py]: The string interpolation should be outside of the translation - you don't want to translate the interpolation. `_('text %s') % var`

File: medical_prescription_state_verify/tests/test_medical_prescription_order.py [hunk 13]
```diff
@@ -20,10 +20,15 @@ def setUp(self):
             'name': 'Test Physician',
             'specialty_id': specialty_id.id,
         })
+        prescription_order_line_id = self.env.ref(
```
Comment (lasley) [inline medical_prescription_state_verify/tests/test_medical_prescription_order.py:23]: remove _id suffix, this is a recordset

File: medical_prescription_state_verify/tests/test_medical_prescription_order.py [hunk 14]
```diff
@@ -41,36 +46,32 @@ def test_allowed_change_states_is_list(self):
 
     def test_write_attrs_not_allowed_when_verified(self):
         record_id = self._new_record()
-        record_id.state_type = 'verified'
+        record_id.write({'stage_id': 4})  # verified
```
Comment (lasley) [inline medical_prescription_state_verify/tests/test_medical_prescription_order.py]: Hard coded database Ids should be replaced for XML ref

File: medical_prescription_state_verify/tests/test_medical_prescription_order_line.py [hunk 15]
```diff
@@ -47,13 +48,13 @@ def _new_record(self):
 
     def test_write_not_allowed_when_verified(self):
         record_id = self._new_record()
-        self.order_id.state_type = 'verified'
+        self.order_id.stage_id = 4  # verified
```
Comment (lasley) [inline medical_prescription_state_verify/tests/test_medical_prescription_order_line.py]: Hard coded db ids

File: medical_prescription_state_verify/models/medical_prescription_order_state.py [hunk 16]
```diff
@@ -1,34 +0,0 @@
...
-    """
-    Add generic types to `medical.prescription.order` for validations
-
-    This class provides generic types to `medical.prescription.order` so that
-    validations and auditing can be applied based on status definitions.
-    """
-
-    _inherit = 'medical.prescription.order.state'
-
-    type = fields.Selection([
```
Comment (lasley) [inline medical_prescription_state_verify/models/medical_prescription_order_state.py]: Why is this removed?

File: sale_crm_medical_prescription/tests/test_crm_lead.py [hunk 17]
```diff
@@ -24,21 +24,24 @@ def setUp(self):
...
         """ Test patient ids properly calculated """
         patient = self.env.ref(
             'sale_crm_medical_prescription.'
             'medical_patient_patient_10'
         )
-        self.assertEquals(
-            [patient.id],
-            self.crm_lead_1.patient_ids.ids,
-        )
+        for id in [patient.id]:
```
Comment (lasley) [inline sale_crm_medical_prescription/tests/test_crm_lead.py]: I think this can be simplified

**Review by lasley — CHANGES_REQUESTED** [review]

**Comment by lasley** [comment]
@bmw95 - please take this over

**Comment by obulkin** [comment]
This PR has been replaced by #213, so I'm closing it.

**Pull request closed without merging**